src/ir/ir_context.h

```diff
@@ -33,29 +37,77 @@ class IRContext {
 
   // Register an operator and return the stored operator instances.
   const Operator &RegisterOperator(std::unique_ptr<Operator> op) {
-    auto ins_res = operators_.insert({std::string(op->name()), std::move(op)});
-    CHECK(ins_res.second) << "Cannot register duplicate operator with name '"
-                          << ins_res.first->first << "'.";
-    return *ins_res.first->second.get();
+    return operators_.RegisterNode(std::move(op), "operator");
   }
 
   // Returns the Operator with a particular name. If there is no
-  // operator with that name, returns nullptr.
+  // operator with that name, fail.
   const Operator &GetOperator(absl::string_view operators_name) const {
-    auto find_res = operators_.find(operators_name);
-    CHECK(find_res != operators_.end())
-        << "Looking up an unregistered operator '" << operators_name << "'.";
-    return *find_res->second;
+    return operators_.GetNode(operators_name, "operator");
   }
 
   // Returns true if the operator is registered with this context.
   bool IsRegisteredOperator(absl::string_view operators_name) const {
-    return operators_.find(operators_name) != operators_.end();
+    return operators_.IsRegisteredNode(operators_name);
+  }
+
+    // Register an operator and return the stored operator instances.
+  const Storage &RegisterStorage(std::unique_ptr<Storage> op) {
+    return storages_.RegisterNode(std::move(op), "storage");
+  }
+
+  // Returns the Storage with a particular name. If there is no
+  // operator with that name, fail.
+  const Storage &GetStorage(absl::string_view operators_name) const {
+    return storages_.GetNode(operators_name, "storage");
+  }
+
+  // Returns true if the operator is registered with this context.
+  bool IsRegisteredStorage(absl::string_view operators_name) const {
+    return storages_.IsRegisteredNode(operators_name);
   }
 
  private:
+  template<class NamedIrNode>
+  class NamedIrNodeRegistry {
```

Comment:
I know `Ir` is according to style guide. Nevertheless, let us stick with `IR` as the capitalization if `IR` is used in class/function names? 

I think we already have `IRContext`. If you strongly prefer `Ir` then we should also change `IrContext`. Let us sync up more on this.

Reply:
I don't have strong preferences, I'll use `IR`.

Reply:
Done.